star/private/timestep.f90

```diff
@@ -495,226 +480,203 @@ integer function check_diffusion_iters_limit(s, skip_hard_limit, dt, dt_limit_ra
+                        max_dX_div_X = delta_X_div_X
+                        max_dX_div_X_j = j
+                        max_dX_div_X_k = k
+                        max_dX_div_X_bdy_dist_dm = bdy_dist_dm
+                     end if
                   end if
-               end if
+               end do
             end do
-         end do
-
-         if (dX_limit > 0) then
-            dX_dt_limit_ratio = max_dX/dX_limit
-            if (dX_dt_limit_ratio <= 1d0) then
-               dX_dt_limit_ratio = 0
-            else
-               j = max_dX_j
-               k = max_dX_k
-               s% Tlim_dX_species = j
-               s% Tlim_dX_cell = k
-               write(*, '(a30, i5, 99e20.10)') &
-                  'dX ' // trim(chem_isos% name(s% chem_id(j))), &
-                  k, max_dX, dX_limit, (s% M_center + s% xmstar*(s% q(k) - s% dq(k)/2))/Msun, &
-                  max_dX_bdy_dist_dm/Msun
+            
+            if (dX_limit(i) > 0) then
+               dX_dt_limit_ratio = max_dX/dX_limit(i)
+               if (ratio_tmp > dX_dt_limit_ratio) then
+                  dX_dt_limit_ratio = ratio_tmp
+                  if (dX_dt_limit_ratio <= 1d0) then
+                     dX_dt_limit_ratio = 0
+                  else
+                     s% Tlim_dX_species = max_dX_j
+                     s% Tlim_dX_cell = max_dX_k
+   !                  write(*, '(a30, i5, 99e20.10)') &
+   !                     ' limit dt because of large dX '// &
+   !                        trim(chem_isos% name(s% chem_id(max_dX_j))) // &
+   !                     ' k, max, lim, m ', &
+   !                     max_dX_k, max_dX, dX_limit(i), &
+   !                     max_dX_bdy_dist_dm/Msun
+                  end if
+               end if
             end if
-
-         end if
-
-         if (dX_div_X_limit > 0) then
-            dX_div_X_dt_limit_ratio = max_dX_div_X/dX_div_X_limit
-            if (dX_div_X_dt_limit_ratio <= 1d0) then
-               dX_div_X_dt_limit_ratio = 0
-            else
-               s% Tlim_dX_div_X_species = max_dX_div_X_j
-               s% Tlim_dX_div_X_cell = max_dX_div_X_k
-               if (which > 2) write(*, '(a30, i5, 99e20.10)') &
-                  'limit dt because of large dX_div_X ' // &
-                     trim(chem_isos% name(s% chem_id(max_dX_div_X_j))) // &
-                     ' k, max, lim, m ', &
-                  max_dX_div_X_k, max_dX_div_X, dX_div_X_limit, &
-                  max_dX_div_X_bdy_dist_dm/Msun
+   
+            if (dX_div_X_limit(i) > 0) then
+               ratio_tmp = max_dX_div_X/dX_div_X_limit(i)
+               if (ratio_tmp > dX_div_X_dt_limit_ratio) then  ! pick out largest culprit only!
+                  dX_div_X_dt_limit_ratio = ratio_tmp
+                  if (dX_div_X_dt_limit_ratio <= 1d0) then
+                     dX_div_X_dt_limit_ratio = 0
+                  else
+                     s% Tlim_dX_div_X_species = max_dX_div_X_j
```

Comment:
@evbauer this property of star is still set, so it _should_ be possible to map that back to the species. As stated though, I think the printing of the string should happen elsewhere, not in this routine.